src/script/api/script_text.cpp

```diff
@@ -167,6 +167,13 @@ const std::string ScriptText::GetEncodedText()
 
 char *ScriptText::_GetEncodedText(char *p, char *lastofp, int &param_count)
 {
+	/* A static vector to prevent circular references. */
+	static std::vector<StringID> seen_ids;
+
+	if (std::find(seen_ids.begin(), seen_ids.end(), this->string) != seen_ids.end()) throw Script_FatalError(fmt::format("{}: Circular reference detected", GetGameStringName(seen_ids[0])));
```

Comment:
If an error gets thrown, it does not get removed from `seen_ids`.

So, I think that if you make a GS with a string error, run that (and thus trigger any of the `throw`s in this function), then fix the GS and restart the GS (or the game, not OpenTTD itself), then it probably throws this error when the same string ID is used again even though there is nothing wrong with that GS.

Reply:
Good catch